rust/dns-over-tcp/src/server.rs

```diff
@@ -0,0 +1,281 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    net::SocketAddr,
+    time::Instant,
+};
+
+use crate::adapter::SmolDeviceAdapter;
+use anyhow::{bail, Context as _, Result};
+use domain::{base::Message, dep::octseq::OctetsInto as _};
+use ip_packet::IpPacket;
+use smoltcp::{
+    iface::{Config, Interface, Route, SocketSet},
+    socket::tcp,
+    storage::RingBuffer,
+    wire::{HardwareAddress, IpEndpoint, Ipv4Address, Ipv4Cidr, Ipv6Address, Ipv6Cidr},
+};
+
+/// A sans-IO implementation of DNS-over-TCP server.
+///
+/// Listens on a specified number of socket addresses, parses incoming DNS queries and allows writing back responses.
+pub struct Server {
+    device: SmolDeviceAdapter,
+    interface: Interface,
+
+    sockets: SocketSet<'static>,
+    listen_endpoints: HashMap<smoltcp::iface::SocketHandle, SocketAddr>,
+
+    received_queries: VecDeque<Query>,
+}
+
+/// Opaque handle to a TCP socket.
+///
+/// This purposely does not implement [`Clone`] or [`Copy`] to make them single-use.
+#[derive(Debug, PartialEq, Eq, Hash)]
+#[must_use = "An active `SocketHandle` means a TCP socket is waiting for a reply somewhere"]
+pub struct SocketHandle(smoltcp::iface::SocketHandle);
+
+pub struct Query {
+    pub message: Message<Vec<u8>>,
+    pub socket: SocketHandle,
+    /// The address of the socket that received the query.
+    pub local: SocketAddr,
+}
+
+const SERVER_IP4_ADDR: Ipv4Address = Ipv4Address::new(127, 0, 0, 1);
+const SERVER_IP6_ADDR: Ipv6Address = Ipv6Address::new(0, 0, 0, 0, 0, 0, 0, 1);
+const NUM_CONCURRENT_SOCKETS: usize = 5;
+
+impl Server {
+    pub fn new(now: Instant) -> Self {
+        let mut device = SmolDeviceAdapter {
+            inbound_packets: VecDeque::default(),
+            outbound_packets: VecDeque::default(),
+        };
+
+        let mut interface =
+            Interface::new(Config::new(HardwareAddress::Ip), &mut device, now.into());
+        // Accept packets with any destination IP, not just our interface.
+        interface.set_any_ip(true);
+
+        // Set our interface IPs. These are just dummies and don't show up anywhere!
+        interface.update_ip_addrs(|ips| {
+            ips.push(Ipv4Cidr::new(SERVER_IP4_ADDR, 32).into()).unwrap();
+            ips.push(Ipv6Cidr::new(SERVER_IP6_ADDR, 128).into())
+                .unwrap();
+        });
+
+        // Configure catch-all routes, meaning all packets given to `smoltcp` will be routed to our interface.
```

Comment:
This aren't catch all routes, right?

Reply:
They are. We only selectively feed packets into `smoltcp` though. This is basically just boilerplate to make sure the `Interface` actually processes all packets that we give to it.